File: api/src/adapters/blobStorage.ts
import { BlobServiceClient, type ContainerClient } from '@azure/storage-blob';
import getSecrets from './keyvault';
import type Image from '../models/image';
import log from '../lib/logger';

let blobServiceClient: BlobServiceClient | null = null;
let imageContainerClient: ContainerClient | null = null;

const createBlobServiceClient = async () => {
  if (blobServiceClient !== null) {
    return blobServiceClient;
  }
  const secrets = await getSecrets();
  const connectionString = secrets.get('blobStorageConnectionString');
  if (connectionString === undefined) {
    throw new Error('Blob connection string is undefined');
  }
  return BlobServiceClient.fromConnectionString(connectionString);
};

const createImageContainerClient = async () => {
  if (imageContainerClient !== null) {
    return imageContainerClient;
  }
  blobServiceClient = await createBlobServiceClient();
  const secrets = await getSecrets();
  const containerName = secrets.get('imageContainerName');
  if (containerName === undefined) {
    throw new Error('Image container name is undefined');
  }
  return blobServiceClient.getContainerClient(containerName);
};

async function streamToText(readable: NodeJS.ReadableStream) {
  readable.setEncoding('utf8');
  let data = '';
  for await (const chunk of readable) {
    data += chunk instanceof Buffer ? chunk.toString() : chunk;
  }
  return data;
}

export const uploadImage = async (image: Image) => {
  const blobName = `${image.name}.${image.ext}`;
  imageContainerClient = await createImageContainerClient();
  const blockBlobClient = imageContainerClient.getBlockBlobClient(blobName);
  log.info({ blobName }, 'Uploading image');
  const uploadResponse = await blockBlobClient.upload(
    image.data,
    image.data.length,
  );
  log.info(
    { blobName, requestId: uploadResponse.requestId },
    'Upload completed',
  );
};

export const downloadImage = async (blobName: string): Promise<Image> => {
  imageContainerClient = await createImageContainerClient();
  const blockBlobClient = imageContainerClient.getBlockBlobClient(blobName);
  log.info({ blobName }, 'Downloading image');
  const downloadBlockBlobResponse = await blockBlobClient.download(0);
  if (downloadBlockBlobResponse.readableStreamBody === undefined) {
    log.error({ blobName }, 'Errod downloading image');
    throw new Error('readableStreamBody is undefined');
  }
  const imagaData = await streamToText(
    downloadBlockBlobResponse.readableStreamBody,
  );
  log.info({ blobName }, 'Downloaded image');
  const splitedBlobName = blobName.split('.');
  return {
    name: splitedBlobName[0],
    ext: splitedBlobName[1],
    data: imagaData,
  };
};

File: backend/adapters/blob_storage.py
from azure.storage.blob import BlobServiceClient

from . import keyvault


def create_blob_service_client():
    secrets = keyvault.get_secrets()
    connection_string = secrets.get('blobStorageConnectionString')
    if connection_string is None:
        raise Exception('Blob connection string is undefined')
    return BlobServiceClient.from_connection_string(connection_string)


def create_image_container_client():
    blob_service_client = create_blob_service_client()
    secrets = keyvault.get_secrets()
    container_name = secrets.get('imageContainerName')
    if container_name is None:
        raise Exception('Image container string is undefined')
    image_container_client = blob_service_client.get_container_client(
        container_name)
    return image_container_client


def download_image(blob_name: str):
    try:
        image_container_client = create_image_container_client()
        block_blob_client = image_container_client.get_blob_client(blob_name)

        download_block_blob_response = block_blob_client.download_blob()
        blob_content = download_block_blob_response.readall()
    except Exception as error:
        return {
            'error': f'Error downloading blob: {str(error)}',
        }
    splited_blob_name = blob_name.split('.')
    return {
        'name': splited_blob_name[0],
        'ext': splited_blob_name[1],
        'data': blob_content,
    }
